### PR TITLE
[WIP] First basic draft of the dispatch system using Redis

### PR DESCRIPTION
**Pull request by lasryaric**
This is a first draft of what the dispatch system could look like, using Redis streams.

The two main Redis commands used are:
- `XADD` to add data to a "persistent stream".
- `XREAD` to read data from the persistant stream.

Let me know what you think.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -13,9 +13,14 @@ services:
   qdrant:
     image: qdrant/qdrant
     volumes:
-      - dustvolume:/qdrant
+      - dustvolume:/qdrant    
     ports:
       - 6334:6334
+      - 6333:6333
```
Comment (spolu) [inline docker-compose.yml:19]: why this change?

**Review by spolu — COMMENTED** [review]
NAICE

**Comment by spolu** [comment]
Reminder that this is on the blocking path for moving to the UI but it looks simple and well advanced enough that we likely will have it ready by tomorrow \o/

**Comment by lasryaric** [comment]
Yes! Will do my best to unlock the API layer asap.

**Comment by lasryaric** [comment]
The new port is dor the http layer of qdrant

**Comment by spolu** [comment]
re qdrant, do we need it?

**Comment by lasryaric** [comment]
Re qdrant port: I use it often for debugging vis curl.

**Comment by spolu** [comment]
Gotcha I realize now this is not impacting production :+1:

**Comment by lasryaric** [comment]
👍 side note but Qdrant cloud exposes this port, and I also use it there for debugging  / profiling sometimes.

File: front/pages/api/v1/w/[wId]/assistant/[cId]/messages/index.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        context: {
+          timezone: payload.context.timezone,
+          username: payload.context.username,
+          fullName: payload.context.fullName,
+          email: payload.context.email,
+          profilePictureUrl: payload.context.profilePictureUrl,
+        },
+      });
+      res.status(200).end();
+      await p;
```
Comment (spolu) [inline front/pages/api/v1/w/[wId]/assistant/[cId]/messages/index.ts:95]: do we need the await? why not void?

File: front/pages/api/v1/w/[wId]/assistant/[cId]/messages/index.ts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    return apiError(req, res, {
+      status_code: 404,
+      api_error: {
+        type: "conversation_not_found",
+        message: "Conversation not found.",
+      },
+    });
+  }
+
+  // no time for actual io-ts parsing right now, so here is the expected structure.
```
Comment (spolu) [inline front/pages/api/v1/w/[wId]/assistant/[cId]/messages/index.ts:59]: roger sounds good :+1:

**Review by spolu — APPROVED** [review]
If conversation sub (which I presume you created for minimal testing) then LGTM

File: front/lib/redis.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import { createClient } from "redis";
+
+let REDIS_CLIENT: ReturnType<typeof createClient> | undefined = undefined;
+
+export async function redisClient() {
```
Comment (spolu) [inline front/lib/redis.ts:5]: What is the env to be setup ?

File: front/lib/api/assistant/pubsub.ts [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+          eventId: messageId,
+          data: payload,
+        };
+      }
+    }
+  }
+}
+
+export async function* getMessagesEvents(
+  messageSID: string,
```
Comment (spolu) [inline front/lib/api/assistant/pubsub.ts:93]: messageId

File: front/lib/api/assistant/pubsub.ts [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+      }
+
+      default:
+        throw new Error(`Unhandled event. ${event.type}`);
+    }
+  }
+}
+
+export async function* getConversationEvents(
+  conversationSID: string,
```
Comment (spolu) [inline front/lib/api/assistant/pubsub.ts:59]: conversationId

File: front/lib/api/assistant/pubsub.ts [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+      case "agent_message_new": {
+        const pubsubChannel = getConversationChannelId(
+          args[1].conversation.sId
+        );
+        await client.xAdd(pubsubChannel, "*", {
+          payload: JSON.stringify(event),
+        });
+        break;
+      }
+      // missing retrieval_documents because it does not have a messageId field.
```
Comment (spolu) [inline front/lib/api/assistant/pubsub.ts:33]: It's been added

File: front/lib/api/assistant/pubsub.ts [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+import {
+  postUserMessage,
+  UserMessageNewEvent,
+} from "@app/lib/api/assistant/conversation";
+import { redisClient } from "@app/lib/redis";
+
+type PostUserMessageParams = Parameters<typeof postUserMessage>;
+
+export async function postUserMessageWithPubSub(
+  ...args: PostUserMessageParams
```
Comment (spolu) [inline front/lib/api/assistant/pubsub.ts:18]: not a big fan of this should we just explicit the parametenrs?

File: front/lib/api/assistant/pubsub.ts [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+  eventId: string;
+  data: UserMessageNewEvent | AgentMessageNewEvent;
+}> {
+  const pubsubChannel = getConversationChannelId(conversationSID);
+  const client = await redisClient();
+
+  while (true) {
+    const events = await client.xRead(
+      { key: pubsubChannel, id: lastEventId ? lastEventId : "0-0" },
+      { COUNT: 1, BLOCK: 10000 }
```
Comment (spolu) [inline front/lib/api/assistant/pubsub.ts:71]: What's the plan with 10s here? what happens once it's reached?

File: front/lib/api/assistant/pubsub.ts [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    | AgentActionEvent
+    | AgentActionSuccessEvent
+    | AgentGenerationTokensEvent
+    | AgentGenerationSuccessEvent;
+}> {
+  const pubsubChannel = getMessageChannelId(messageSID);
+  const client = await redisClient();
+  const events = await client.xRead(
+    { key: pubsubChannel, id: lastEventId ? lastEventId : "0-0" },
+    { COUNT: 1, BLOCK: 10000 }
```
Comment (spolu) [inline front/lib/api/assistant/pubsub.ts:108]: We need at least 1mn I think?

**Review by spolu — CHANGES_REQUESTED** [review]
Actually a few comments

**Comment by lasryaric** [comment]
There is a weird behavior that I am not understanding yet with the
subscribe (sub) endpoint.
Give me a bit more time before using it (~15 minutes).

**Comment by lasryaric** [comment]
Ready for review on my side.

File: front/lib/api/assistant/pubsub.ts [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+  data: UserMessageNewEvent | AgentMessageNewEvent;
+}> {
+  const pubsubChannel = getConversationChannelId(conversationId);
+  const client = await redisClient();
+
+  while (true) {
+    const events = await client.xRead(
+      { key: pubsubChannel, id: lastEventId ? lastEventId : "0-0" },
+      // weird, xread does not return on new message when count is = 1. Anything over 1 works.
+      { COUNT: 10, BLOCK: 60 * 1000 }
```
Comment (spolu) [inline front/lib/api/assistant/pubsub.ts:89]: `// weird` right in the middle of the dispatch system is not great.

COUNT must have an impact no? Does it work if we have more than 10 events in the queue?

File: front/pages/api/v1/w/[wId]/assistant/conversation/[cId]/sub.ts [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    return apiError(req, res, {
+      status_code: 400,
+      api_error: {
+        type: "invalid_request_error",
+        message: "The Assistant API is only available on your own workspace.",
+      },
+    });
+  }
+
+  const conv = await Conversation.findOne({
```
Comment (spolu) [inline front/pages/api/v1/w/[wId]/assistant/conversation/[cId]/sub.ts:43]: Let's remove the Conversation loading for now

Comment (spolu) [reply front/pages/api/v1/w/[wId]/assistant/conversation/[cId]/sub.ts:43]: What happens if we call getConversationEvents on a bogus sId?

Comment (lasryaric) [reply front/pages/api/v1/w/[wId]/assistant/conversation/[cId]/sub.ts:43]: it will block until the timeout has been reached out and return 0 event.

File: front/lib/api/assistant/pubsub.ts [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+  data: UserMessageNewEvent | AgentMessageNewEvent;
+}> {
+  const redis = await redisClient();
+  const pubsubChannel = getConversationChannelId(conversationId);
+
+  try {
+    while (true) {
+      const events = await redis.xRead(
+        { key: pubsubChannel, id: lastEventId ? lastEventId : "0-0" },
+        // weird, xread does not return on new message when count is = 1. Anything over 1 works.
```
Comment (philipperolet) [inline front/lib/api/assistant/pubsub.ts:94]: Hey! Reading the code to prepare, thanks for all that aric :pray: 

NRE - Just asking : something I don't understand here: if we use count = 1 then only 1 event on the array max right, and no need for the loop below? Why not get e.g. count = 100, to limit the calls to redis -- and keep the loop?

Comment (philipperolet) [reply front/lib/api/assistant/pubsub.ts:94]: Sorry to drop in uninvited :sweat_smile: hope I'm not disturbing

Comment (lasryaric) [reply front/lib/api/assistant/pubsub.ts:94]: For context: this comment was wrong, count is honored, I am not sure exactly if redis can return more than 1 event when using the blocking call because I have read that it returns on the first one, but for the first call with no "lastEventId" set, it does return multiple rows.

In any case, that's the redis interface, they always return an array, so we'll always have to loop. We could fetch event[0] but it does not change the code much I think. Does it?
